### ISSUE TITLE
should Spring beans explicitly declare "prototype" scope

### ISSUE THREAD

**Issue by gavinking** [samolisov/spring-vs-ejb-vs-cdi-benchmark#1]
I'm not certain, but I think Spring beans are singletons by default. So to be fair, the benchmark should explicitly assign them to prototype scope.

**Comment by olivergierke** [comment]
You're right, they are. But I don't think they should. The use case absolutely doesn't need any component  to be prototype scoped and if you can avoid re-creation of objects why would you want to actively avoid the avoidance of overhead? If other technologies decide to default to a scope that's sub-optimal for this particular case, I think the example could be tweaked to make sure e.g. the CDI parts use application scoped beans.

The great thing about your suggestion though is that it points to a fundamental difference in the way the containers handle component scopes (one of the reasons Spring never would/could actually implement CDI without turning its container model upside down) due to the history and context they were developed in. In a Spring application defining a different scope than singleton is only used occasionally as application components are considered thread-safe by default and scoped instances (request-bound information etc.) is rather passed down via method arguments rather than component injection points (which inevitably requires the containing component to be recreated and thus being scoped accordingly).

In a CDI world, scoping is a much more natural concept, which is probably coming from its history in Seam and thus JSF. That of course makes it feel more aligned if used in that scenario but also comes with the overhead or additional configuration effort in simple cases like a REST resource and thread-safe downstream components. Exactly the other way round for Spring. Simpler in that case, but requires more effort in terms of scoping in a context heavily relying on scoping.

Long story short, Spring works as expected for the use case the benchmark showcases. Moving to prototype scoped beans would distort that, as nobody building a real Spring application would make all her beans prototype scoped. Moving the CDI variant to application scoped beans — all in. 🙃

**Comment by gavinking** [comment]
Well, OK, but then it's totally unclear what your benchmark is showing. Is it showing a real difference in performance between the containers, or is it just showing the expected result of a difference between component lifecycles? It makes the benchmark impossible to interpret.

**Comment by olivergierke** [comment]
It's not my benchmark 🙃. I just saw it flying through my Twitter stream. Totally agree that the benchmark is flawed as it includes so many things except the allegedly to compare tech stacks that can influence the numbers drastically. Plus, Spring applications are usually deployed to plain servlet containers, which complicates the comparison even more. I haven't even run it myself as I luckily happen to not have a WebSphere around we usually just bootstrap embedded containers these days.

To be completely honest, I think the entire idea here is flawed as it would be surprising if the numbers were totally out of the ballpark for one of the competitors. If they really were, I guess that could be fixed easily. Plus, I think whatever "performance" means in real applications is not tied to the time spend during injection but in database access, response rendering, network latency etc.

**Comment by johnament** [comment]
Yes, but for this benchmark to be fair you (the author) needs to test apples to apples.  Things like constructor injection of path params logically needs request/prototype scoping in place or else the values are cached.  It would be good to have two sets of tests

- Spring prototype vs CDI request scope vs EJB SLSB
- Spring singleton vs CDI app scope vs EJB Singleton

Rafael's taking care of one part, but I suppose adding a second set of tests would help as well.
